android-app/app/src/main/java/de/dhbw/ka/se/fibo/utils/backend/BackendSynchronizationFactory.java

```diff
@@ -0,0 +1,154 @@
+        if (!isAuthenticated) {
+            throw new IllegalStateException("need to have an access token to sync!");
+        }
+
+        Optional<String> accessToken = ApplicationState.getInstance(context).getAccessToken();
+
+        Runnable r = () -> {
+            SharedVolleyRequestQueue requestQueue = SharedVolleyRequestQueue.getInstance(context);
+            CountDownLatch latch = new CountDownLatch(3);
+
+            // build requests
+            BackendCashflowListRequest cashflowsRequest = new BackendCashflowListRequest(latch, accessToken.get());
+            BackendCategoryListRequest categoriesRequest = new BackendCategoryListRequest(latch, accessToken.get());
+            BackendPlacesListRequest placesRequest = new BackendPlacesListRequest(latch, accessToken.get());
+
+            // enqueue requests
+            requestQueue.addToRequestQueue(cashflowsRequest.getRequest());
+            requestQueue.addToRequestQueue(categoriesRequest.getRequest());
+            requestQueue.addToRequestQueue(placesRequest.getRequest());
+
+            // wait for all requests to finish
+            try {
+                latch.await();
+            } catch (InterruptedException e) {
+                failWithThrowable(e);
+                return;
+            }
+
+            // handle responses
+            try {
+                BackendSynchronizationResult result = mergeResponses(cashflowsRequest.getResponse(), categoriesRequest.getResponse(), placesRequest.getResponse());
+                propagateResult(result);
+            } catch (VolleyError e) {
+                failWithThrowable(e);
+            }
+        };
+
+        // waiting for the latch is blocking, hence we run this on another thread
+        new Thread(r).start();
+    }
+
+    private BackendSynchronizationResult mergeResponses(CashflowListResponse[] cashflows, CategoryListResponse[] categories, PlacesListResponse[] places) {
+        Map<Integer, Category> categoryLookupTable = new HashMap<>();
+        Map<Integer, Place> placesLookupTable = new HashMap<>();
+        SortedSet<Cashflow> newCashflowSet = new TreeSet<>();
+
+        for (CategoryListResponse serializedCategory : categories) {
+            categoryLookupTable.put(serializedCategory.id, serializedCategory.toCategory());
+        }
+
+        for (PlacesListResponse serializedPlace : places) {
+            placesLookupTable.put(serializedPlace.id, serializedPlace.toPlace());
+        }
+
+        for (CashflowListResponse serializedCashflow : cashflows) {
+            Category category = categoryLookupTable.get(serializedCashflow.categoryID);
+            Place place = placesLookupTable.get(serializedCashflow.placeID);
+
+            Cashflow cashflow = new Cashflow(category, serializedCashflow.getCashflowType(), serializedCashflow.overallValue, serializedCashflow.created, place);
+
+            newCashflowSet.add(cashflow);
+        }
```

Comment:
looks for me like the `.map()` method could be used instead of the `for each` loops here, doesn't it?

Reply:
It could be used, it's personal preference. I think foreach loops are more explicit and easier to understand.

Would you like to see this changed?

Reply:
No you dont have to implement it. I did not try whether if works the way I imagined and you can do whatever fits more for your style. It was just a hint :)